envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoySnapshotFactory.kt

```diff
@@ -115,21 +154,30 @@ internal class EnvoySnapshotFactory(
             it.tags.contains(properties.egress.http2.tagName)
         }
 
-        return ClusterConfiguration(serviceName, allInstancesHaveEnvoyTag)
+        val http2Status = if (allInstancesHaveEnvoyTag) {
+            Http2Status.ENABLED
+        } else if (allInstances.isNotEmpty()) {
+            Http2Status.DISABLED
+        } else {
+            // decide later based on previous cluster config
```

Comment:
I do not like deciding later, this seems convoluted and makes you jump around the code. Can't we just pass previous state to this function and decide here?

Reply:
My main argument against is that we would have to compute `clusterConfiguration` logic twice for every state. First time, when the state is acting as "current" state, and second time when the same state is acting as "previous" state. This PR was created as an optimisation as a main goal, so I focused on performance. But we can talk about this.

I think with enum we can jump fairly easy to see how UNKNOWN state is handled (if we use IDE, not so easy when we review PR in Github)

Reply:
@lukidzi @pbetkier WDYT?

Reply:
I agree with @slonka that current implementation is not super readable. I don't understand the problem you describe @MarcinFalkowski. If accessing previous state would require unnecessary recomputation how come we have it later in `modifyClustersUsingPreviousState`? I don't really get the flow :)

Reply:
@pbetkier this code would be executed twice for every state:

```kotlin
    fun clusterConfigurations(servicesStates: List<LocalityAwareServicesState>): Map<String, ClusterConfiguration> {
        return if (properties.egress.http2.enabled) {
            servicesStates.flatMap {
                it.servicesState.serviceNameToInstances.values
            }.groupBy {
                it.serviceName
            }.mapValues { (serviceName, instances) ->
                toClusterConfiguration(instances, serviceName)
            }
        } else {
            servicesStates
                .flatMap { it.servicesState.serviceNames() }
                .distinct()
                .associateWith { ClusterConfiguration(serviceName = it, http2 = Http2Status.DISABLED) }
        }
    }
```

Right now we compute it once and then use it in `Flux.scan` 

However, as you noticed, if we change it, we can remove `modifyClustersUsingPreviousState`: 
```
    fun modifyClustersUsingPreviousState(
        previous: Map<String, ClusterConfiguration>,
        current: Map<String, ClusterConfiguration>
    ): Map<String, ClusterConfiguration> {

        val anyRemoved = if (properties.egress.neverRemoveClusters) {
            previous.keys.any { it !in current }
        } else {
            false
        }
        val anyUnknownHttp2 = current.any { it.value.http2 == Http2Status.UNKNOWN }

        if (!anyRemoved && !anyUnknownHttp2) {
            // fast path (most frequent)
            return current
        }

        val fixedClusters = when (anyUnknownHttp2) {
            true -> setHttp2StatusUsingPreviousState(previous, current)
            false -> current
        }
        return when (anyRemoved) {
            true -> {
                val removedClusters = previous - current.keys
                fixedClusters + removedClusters
            }
            false -> fixedClusters
        }
    }
```

But this code is far more efficient, because there is no grouping and flat mapping, and for most cases it will do almost nothing (see `fast path` comment)

Reply:
OK, I fixed the flow without a penalty to a performance. Thank @slonka for forcing me to rethink it!